File: GAMEOFLIFE/Animal.cpp
#include <ncurses.h>
using namespace std;

#include "Animal.h"
#include "World.h"

Animal::Animal(char icon, int colornum, int x, int y, int strength, int initiative) : Organism(icon, colornum, x, y, strength, initiative) {
}

void Animal::Action() {

	std::vector<Organism*> organisms = world->GetOrganisms();

	AddAge();

    std::random_device rd;
    std::mt19937 eng(rd());
    std::uniform_int_distribution<> range(0, 3);

	int move = range(eng);
	
	while (!SetPos({ GetX()+moves[move][0], GetY()+moves[move][1] }, false)) {
		move++;
		move %= 4;
	}
}

void Animal::Collision(Organism *other)
{
	if(this==other){
		return;
	}
	if (typeid(*this) == typeid(*other)) {


		if(GetAge()<15||other->GetAge()<15){
			return;
		}

		auto child = dynamic_cast<Animal*>(this->Child());

		if (child != nullptr) {
			std::random_device rd;
			std::mt19937 eng(rd());
			std::uniform_int_distribution<> range(0, 3);

			int pick = range(eng);

			while (!child->SetPos({GetX()+moves[pick%4][0], GetY()+moves[pick%4][1]}, true)) {

				pick++;

				if (pick > 8) {
					world->AddLog(this, "no space for child", 1);
					child->Kill();
					return;
				}
			}
			world->AddLog(this, "baby is born", 1);
				
		}
	}
	else if (RunAway() || other->RunAway()) {
		return;
	}
	else if (GetStrength() > other->GetStrength()) {

		world->AddLog(this, "perseveres", 1);
		other->Kill();
	}
	else if (GetStrength() < other->GetStrength()) {

		world->AddLog(this, "dies", 1);
		Kill();
	}
	else {
		world->AddLog(this, "perseveres", 1);
		other->Kill();
	}
}

File: GAMEOFLIFEJAVA/gameoflife/organisms/animals/Animal.java
package gameoflife.organisms.animals;

import gameoflife.structs.Vector;
import gameoflife.structs.RandomGenerator;
import gameoflife.organisms.Organism;

public class Animal extends Organism {

    public Animal(String icon, int strength, int initiative, Vector position) {
        super(icon, strength, initiative, position);
    }

    public void Action() {

        RandomGenerator r = new RandomGenerator();
        AddAge();

        Vector[] allowedMoves = world.GetPossibleMoves();
        int move = r.random(0,allowedMoves.length-1);

        while(!SetPos(GetPos().Add(allowedMoves[move]))){
            move++;
            move%=allowedMoves.length;
        }
    }

    public void Collision(Organism other) {

        if(this.getClass().equals(other.getClass())){
            if(this.GetAge()<10||other.GetAge()<10){
                return;
            }
            try {
                Vector[] allowedMoves = world.GetPossibleMoves();
                int move = 0;
                while(!world.OnBoard(GetPos().Add(allowedMoves[move]))||world.GetOrganism(GetPos().Add(allowedMoves[move]))!=null){
                    move++;
                    if(move==allowedMoves.length) {
                        world.Log("No place for baby " + this.getClass().getSimpleName());
                        return;
                    }
                }
                RevertMove();
                world.AddOrganism(this.getClass().getCanonicalName(), GetPos().Add(allowedMoves[move]));
            } catch (Exception e) {
                System.err.println(e.getMessage());
            }
            world.Log(this.getClass().getSimpleName()+" baby has been born");
            return;
        }
        if(this.Escaped()||other.Escaped()) {
            return;
        }
        if(this.GetStrength()>=other.GetStrength()){
            world.Log(this.getClass().getSimpleName()+" killed "+other.getClass().getSimpleName()+"!");
            other.Kill();
        } else {
            world.Log(this.getClass().getSimpleName()+" has been killed by "+other.getClass().getSimpleName()+"!");
            this.Kill();
        }
    }
    
}
